buidl/test/test_musig.py

```diff
@@ -277,45 +570,84 @@ def test_multi_leaf_multisig_sequence(self):
         tap_root_input = tr_multisig.multi_leaf_tap_root(sequence=current_sequence)
-        tap_root_output = tr_multisig.musig_tap_root(locktime=Locktime(1638000000))
-        leaf = MultiSigTapScript(points[:3], 3, sequence=current_sequence).tap_leaf()
         self.assertEqual(
             tap_root_input.address(network="signet"),
-            "tb1px58s95zf3z7nmjl6d7p46lczgp52mz8z3lg3a4ck7grf0je0pl6q5gqdqu",
+            "tb1p2qgnkwl2ny40d0xcy0fldljzmsk5vpj8navrhh5y2vsvle6xnj5ssvzrck",
         )
         prev_tx = bytes.fromhex(
-            "cf31d8a1f12c80407a2e6221f696a9f5300fe9c72fbdd3b6dd1b0a779b9da28f"
-        )
-        fee = 204
-        tx_in = TxIn(prev_tx, 0, sequence=current_sequence)
-        tx_in._value = 9999396
-        tx_in._script_pubkey = tap_root_input.script_pubkey()
-        tx_out = TxOut(tx_in.value() - fee, tap_root_output.script_pubkey())
-        tx_obj = Tx(2, [tx_in], [tx_out], 0, network="signet", segwit=True)
-        tx_in.witness.items = []
-        tx_obj.initialize_p2tr_multisig(
-            0,
-            tap_root_input.control_block(leaf),
-            leaf.tap_script,
-        )
-        sigs = [
-            tx_obj.get_sig_taproot(0, priv, ext_flag=1) for priv in private_keys[:3]
-        ]
-        tx_obj.finalize_p2tr_multisig(0, sigs)
-        self.assertEqual(fee, tx_obj.vbytes())
+            "40a61c5dd4a8c84e52563e5a47c6f962fa423b07785a1b412174eaef4df64b7b"
+        )
+        tx_ins = []
+        for prev_index in range(11):
+            tx_in = TxIn(prev_tx, prev_index, sequence=current_sequence)
+            tx_in._value = 1000000
+            tx_in._script_pubkey = tap_root_input.script_pubkey()
+            tx_ins.append(tx_in)
+        change_script = address_to_script_pubkey(
+            "tb1q7kn55vf3mmd40gyj46r245lw87dc6us5n50lrg"
+        )
+        fee = 1629
+        tx_out = TxOut(1000000 * len(tx_ins) - fee, change_script)
+        tx_obj = Tx(2, tx_ins, [tx_out], 0, network="signet", segwit=True)
+        for input_index, pubkeys in enumerate(combinations(points, 3)):
+            leaf = MultiSigTapScript(pubkeys, 3, sequence=current_sequence).tap_leaf()
+            self.assertTrue(tap_root_input.control_block(leaf))
+            tx_in = tx_ins[input_index]
+            tx_in.witness.items = []
+            tx_obj.initialize_p2tr_multisig(
+                input_index,
+                tap_root_input.control_block(leaf),
+                leaf.tap_script,
+            )
+            sigs = []
+            for priv in private_keys:
+                if priv.point in pubkeys:
+                    sig = tx_obj.get_sig_taproot(input_index, priv, ext_flag=1)
+                else:
+                    sig = b""
+                sigs.append(sig)
+            tx_obj.finalize_p2tr_multisig(input_index, sigs)
+            self.assertTrue(tx_obj.verify_input(input_index))
+        input_index = len(tx_ins) - 1
+        tx_in = tx_ins[input_index]
+        musig = MuSigTapScript(points)
```

Comment:
Ditto the below, should this have a helper method? Seems not only very advanced for the caller but something where you want to centralize the logic in a place that has nice test coverage.

Reply:
We're simulating each key signer generating two nonces for the MuSig2 protocol. This is where you have multiple signers produce their own nonces so it doesn't make much sense to have a function that generates all of them at once as that shouldn't be happening. I'm also against a helper function within the test suite because that tends to make the tests less readable and one change tends to break lots of tests. Also, those functions tend to get very complicated as just one more condition to add yet another test case makes it really hard to debug. I generally just like having test cases be really readable and not call other stuff that's not in the library.